crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,479 @@
+//! Defines the [`AssetChanged`] query filter.
+//!
+//! Like [`Changed`](bevy_ecs::prelude::Changed), but for [`Asset`]s,
+//! and triggers whenever the handle or the underlying asset changes.
+
+use crate::{AsAssetId, Asset, AssetId};
+use bevy_ecs::component::Components;
+use bevy_ecs::{
+    archetype::Archetype,
+    component::{ComponentId, Tick},
+    prelude::{Entity, Resource, World},
+    query::{FilteredAccess, QueryFilter, QueryItem, ReadFetch, WorldQuery},
+    storage::{Table, TableRow},
+    world::unsafe_world_cell::UnsafeWorldCell,
+};
+use bevy_utils::tracing::error;
+use bevy_utils::HashMap;
+use core::marker::PhantomData;
+use disqualified::ShortName;
+
+/// A resource that stores the last tick an asset was changed. This is used by
+/// the [`AssetChanged`] filter to determine if an asset has changed since the last time
+/// a query ran.
+///
+/// This resource is automatically managed by the [`AssetEvents`](crate::AssetEvents) schedule and
+/// should not be exposed to the user in order to maintain safety guarantees. Any additional uses of
+/// this resource should be carefully audited to ensure that they do not introduce any safety
+/// issues.
```

Comment:
When you mean "should not be exposed", what do you mean? Changing the fields?

I ask because usually safety invariants are scoped to be module-private, not crate-private.

Reply:
This mostly is meant to mean don't make this pub because some asks. It's crate scoped mostly due to code organization, having the filter live in its own file is I think preferable to including all of this code in the root lib file where the asset event system lives simply for the sake of keeping it module private.